src/cbor_decode.c

```diff
@@ -0,0 +1,390 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <string.h>
+#include "cbor_decode.h"
+
+static uint32_t additional_len(uint8_t additional)
+{
+	switch (additional) {
+	case 24: return 1;
+	case 25: return 2;
+	case 26: return 4;
+	case 27: return 8;
+	default: return 0;
+	}
+}
+
+#define COMP_PTRS(type, p_res, p_min, p_max) \
+		(((p_min == NULL) || (*(type *)p_res >= *(type *)p_min)) \
+		&& ((p_max == NULL) || (*(type *)p_res <= *(type *)p_max)))
+
+static bool value_extract(uint8_t const **const pp_payload,
```

Comment:
This function *may* fail (intentionally) if this is used in a "try parse" (e.g. deserialization of a union when you don't know the content)

Suggest a comment around this function to explain that in certain cases returning false does not break down the decoding, and in this case the read-pointer gets reset automatically by this function

Also suggested is to create a special chapter inside documentation showing the process of decoding unions (as this is something not every serialization scheme has support for)

Reply:
This is now explained in a few sentences in the file docs in cbor_decode.h